Channels

Erykiusz opened the pull request.

jamesmkrieger committed:
fix links in readme

jamesmkrieger committed:
Merge branch 'main' of github.com:prody/ProDy into readme

karolamik13 committed:
Merge pull request #1925 from jamesmkrieger/readme

fix links in readme

Erykiusz committed:
Work in progress channel detection module

Erykiusz committed:
new commit

Erykiusz committed:
new commit

karolamik13 commented:
Please move non-standard imports inside the functions and use LOGGER to check whether the package is installed. If not, provide information on which particular package needs to be installed to use this function. If we have those imports in the current form, I immediately have an error when importing ProDy (see below).

In [1]: from prody import *
---------------------------------------------------------------------------
ModuleNotFoundError                       Traceback (most recent call last)
Cell In[1], line 1
----> 1 from prody import *

File ~/test_ProDy/ET_ProDy/ProDy/prody/__init__.py:88
     84 __all__.append('atomic')
     86 from .atomic import SELECT
---> 88 from . import proteins
     89 from .proteins import *
     90 __all__.extend(proteins.__all__)

File ~/test_ProDy/ET_ProDy/ProDy/prody/proteins/__init__.py:253
    250 from .starfile import *
    251 __all__.extend(starfile.__all__)
--> 253 from . import interactions
    254 from .interactions import *
    255 __all__.extend(interactions.__all__)

File ~/test_ProDy/ET_ProDy/ProDy/prody/proteins/interactions.py:37
     35 import subprocess
     36 import heapq
---> 37 import open3d as o3d
     38 from collections import deque
     39 from scipy.interpolate import CubicSpline

ModuleNotFoundError: No module named 'open3d'

karolamik13 commented:
I would suggest changing the name of the main function detect_channels to calcChannels. ProDy is using calcXXX (to compute something) or getXXX (if we want raw numbers, etc.).

Here:
detect_channels(
    **protein,**
    r1=3,
    r2=1.25,
    min_depth=10,
    bottleneck=1,
    sparsity=15,
    output_path=None,
    visualizer=None,
    stl_path=None,
)

in Prody we are using "atoms" as the parameter with protein structure.

See here:
:arg atoms: an Atomic object from which residues are selected
:type atoms: :class:`.Atomic`

karolamik13 commented:
Usually, in ProDy, some results are returned. See below...when I am calling channels, I should have some results on which I could do some further analysis.. for example, to check the volume of the channels or which residues are forming the channel, etc. We should be able to have access to each channel among those 9 which were identified. Dictionary maybe?


In [10]: p = parsePDB('1tqn.pdb')
@> 3999 atoms and 1 coordinate set(s) were parsed in 0.12s.

In [11]: p_prot = p.select('protein')

In [12]: p_prot
Out[12]: <Selection: 'protein' from 1tqn (3766 atoms)>

In [13]: channels = detect_channels(p_prot)
9
Detected 9 channels.
No output path given.

In [14]: channels

karolamik13 commented:
visualizer='channels' - that looks nice. 
Can we have another option that if we set it to visualizer='True', we will see the channels + protein in the NewCartoon representation?

karolamik13 reviewed (commented):
Except for the comments above, the channel prediction looks good. I am receiving similar results. When the comments will be applied I will test it again on different structures and with different parameters.

Erykiusz commented:
The stl file is created from the run_vmd_script command or the tcl script itself. 'stl_path' is an input parameter with a 3D model of the NewCartoon representation of the protein. When the parameter is set to None, instead of the protein, a triangle mesh approximating the protein surface will be used. When the program detects an stl file with a protein model, it automatically displays the protein instead of the approximated mesh.

Erykiusz committed:
added output for main function and fixed some features

Erykiusz commented:
New changes:
-added new function getChannelsParameters which provides details about lengths and bottlenecks of channels
-changed names of functions: runVmdScript and calcChannels
-fixed runVmdScript so it now works
-moved libriaries inside the function so they can be chekked with logger
-calcChannels now returns channels in a list object, also returns surface of a molecule. Added two new parameters: 'vis_channels' and 'surface', enabling to use the function in only visualizer mode. If 'vis_channels' parameter is not set to None, only visualization of channels inputed in 'vis_channels' will be performed. This enables to visualize only channels of interest. Example use:

protein_path = "C:\\Users\\User\\Desktop\\1tqn.pdb"
vmd_path = "C:\\Program Files\\VMD\\vmd.exe"
script_path = "C:\\Users\\User\\vmd_script.tcl"
stl_path = "C:\\Users\\User\\Desktop\\result.stl" #this file will be created in runVmdScript
prody.runVmdScript(vmd_path, protein_path, script_path, stl_path)

protein = prody.parsePDB(protein_path) 
channels, surface = prody.calcChannels(protein) #calculations
prody.calcChannels(protein, vis_channels = channels, surface = surface, stl_path = stl_path) #visualization only
lengths, bottlenecks = prody.getChannelsParameters(channels)

I didnt yet add volume of channels in getChannelsParameters because it will take a bit longer to implement.

karolamik13 commented:
Thank you for the improvements. 

Now I am testing the functions again:

In [9]: channels = calcChannels(atoms)
Detected 10 channels.
No output path given.

In [10]: channels
Out[10]: 
([<prody.proteins.interactions.Channel at 0x7fc51f663cd0>,
  <prody.proteins.interactions.Channel at 0x7fc51f663bb0>,
  <prody.proteins.interactions.Channel at 0x7fc51f663af0>,
  <prody.proteins.interactions.Channel at 0x7fc51f663a30>,
  <prody.proteins.interactions.Channel at 0x7fc51f663970>,
  <prody.proteins.interactions.Channel at 0x7fc51f663850>,
  <prody.proteins.interactions.Channel at 0x7fc51f6635b0>,
  <prody.proteins.interactions.Channel at 0x7fc51f663430>,
  <prody.proteins.interactions.Channel at 0x7fc51f663370>,
  <prody.proteins.interactions.Channel at 0x7fc51f6632b0>],
 array([[5136, 5139, 5140, 5137],
        [5138, 5136, 5139, 5137],
        [5138, 5139, 5140, 5137],
        ...,
        [1163, 3246, 3249, 3248],
        [1163, 1164, 3246, 3249],
        [1163, 3246, 3284, 3248]], dtype=int32))


It's nice that we have some results now. Can we somehow distinguish different channels? We have 10 which were detected in the structure I tested. 

Also, "No output path given." I would change it to LOGGER info. Something like: To obtain a PDB file with the results please use...this and this parameter, etc. (or some other information to explain more what to do to actually obtain some visualization of the channels and not only the number.

Erykiusz commented:
To distinguish between channels we can use getChannelsParameters and check the lengths. Parameters are outputted to a list which order corresponds to the order of channels. For example, if we want to get only the longest channel, we would check the lengths of channels and get an index of the longest channels (for example 3), and then get the 3rd channel by doing channels[3] and input it to the visualizer by prody.calcChannels(protein, vis_channels = channels[3], surface = surface). I dont really have an idea to distinguish the channels in a different way.

karolamik13 commented:
Hi James @jamesmkrieger
this is the first version of the channel detector for ProDy. 
Let us know if you have any suggestions. It is still under development.
In the end, when everything is working fine, we might want to create a class for this kind of calculation and move channels from interaction.py to the separated file.

karolamik13 commented:
@Erykiusz could you provide an example line by line? (even just commands)
I am trying things, but I have errors.

Also, can we have access to the atoms (=channels) that are saved as PDB files? It would be very useful to have coordinates of those atoms and then we can use exwithin in select to provide the residues which are forming the channel. And it would be good to have those atoms separated by channels.

karolamik13 commented:
Now it is working fine. Thank you.

jamesmkrieger commented:
Great. I’ll have a look.

I don’t think it needs to go in a class. We have many functionalities in prody that are just functions too.

jamesmkrieger commented:
> Please move non-standard imports inside the functions and use LOGGER to check whether the package is installed. If not, provide information on which particular package needs to be installed to use this function. If we have those imports in the current form, I immediately have an error when importing ProDy (see below).
> 
> ## In [1]: from prody import *
> ModuleNotFoundError Traceback (most recent call last) Cell In[1], line 1 ----> 1 from prody import *
> 
> File ~/test_ProDy/ET_ProDy/ProDy/prody/**init**.py:88 84 **all**.append('atomic') 86 from .atomic import SELECT ---> 88 from . import proteins 89 from .proteins import * 90 **all**.extend(proteins.**all**)
> 
> File ~/test_ProDy/ET_ProDy/ProDy/prody/proteins/**init**.py:253 250 from .starfile import * 251 **all**.extend(starfile.**all**) --> 253 from . import interactions 254 from .interactions import * 255 **all**.extend(interactions.**all**)
> 
> File ~/test_ProDy/ET_ProDy/ProDy/prody/proteins/interactions.py:37 35 import subprocess 36 import heapq ---> 37 import open3d as o3d 38 from collections import deque 39 from scipy.interpolate import CubicSpline
> 
> ModuleNotFoundError: No module named 'open3d'

I wouldn’t use LOGGER to report modules not installed. We need to raise an ImportError inside the function saying that the user needs that module to run the function

jamesmkrieger commented on prody/proteins/interactions.py:
```diff
@@ -4345,4 +4344,772 @@ def saveInteractionsPDB(self, **kwargs):
+    from scipy.spatial import Voronoi, Delaunay
+    from pathlib import Path
+    
+    if not check_and_import('heapq'):
```
Check packages before trying to import them to not get errors

jamesmkrieger replied:
You can probably also do this with a loop over a list of names

jamesmkrieger replied:
This could also be a general function in prody.utilities and I was thinking of adding import_command inside too

jamesmkrieger replied:
I now have the following code that seems to work:
```
    def check_and_import(package_name, import_command=None):
        """Check for package and import it if possible and return **True**.
        Otherwise, return **False
        
        :arg package_name: name of package
        :type package_name: str

        :arg import_command: optional command to import submodules or with an alias
            default **None** means use "import {0}".format(package_name)
        :type import_command: None, str
        """

        if not isinstance(package_name, str):
            raise TypeError('package_name should be a string')
        
        if not isinstance(import_command, (type(None), str)):
            raise TypeError('import_command should be None or a string')

        import importlib.util
        if importlib.util.find_spec(package_name) is None:
            LOGGER.warn(f"Package '{package_name}' is not installed. Please install it to use this function.")
            return False
        
        if import_command is None:
            exec('import {0}'.format(package_name))
        else:
            exec(import_command)

        return True
    
    required = [['heapq', None], ['open3d', 'import open3d as o3d'], ['collections', 'from collections import deque'],
                ['scipy', 'from scipy.interpolate import CubicSpline; from scipy.spatial import Voronoi, Delaunay'],
                ['pathlib', 'from pathlib import Path']]
    
    missing = []
    errorMsg = None
    for (name, command) in required:
        if not check_and_import(name, command):
            missing.append(name)
            if errorMsg is None:
                errorMsg = 'To run calcChannels, please install {0}'.format(missing[0])
            else:
                errorMsg += ', ' + name

    if len(missing) > 0:
        if len(missing) > 1:
            errorMsg = ', '.join(errorMsg.split(', ')[:-1]) + ' and ' + errorMsg.split(', ')[-1]
        raise ImportError(errorMsg)
```

jamesmkrieger replied:
In fact, we then don't need the logger step at all. The current output is as follows:
```
channels, surface = prody.calcChannels(protein)
@> WARNING Package 'open3d' is not installed. Please install it to use this function.
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
Cell In[54], line 1
----> 1 channels, surface = prody.calcChannels(protein)

File ~/software/scipion3/software/em/prody-github/ProDy/prody/proteins/interactions.py:4511, in calcChannels(atoms, r1, r2, min_depth, bottleneck, sparsity, output_path, visualizer, stl_path, vis_channels, surface)
   4509     if len(missing) > 1:
   4510         errorMsg = ', '.join(errorMsg.split(', ')[:-1]) + ' and ' + errorMsg.split(', ')[-1]
-> 4511     raise ImportError(errorMsg)
   4513 class State:
   4514     def __init__(self, simplices, neighbors, vertices):

ImportError: To run calcChannels, please install open3d
```

jamesmkrieger replied:
ok, it looks like importing inside the check function doesn't work. Sorry.

Now my code is the following:
```
    def check_and_import(package_name):
        """Check for package and import it if possible and return **True**.
        Otherwise, return **False
        
        :arg package_name: name of package
        :type package_name: str

        :arg import_command: optional command to import submodules or with an alias
            default **None** means use "import {0}".format(package_name)
        :type import_command: None, str
        """
        if not isinstance(package_name, str):
            raise TypeError('package_name should be a string')

        import importlib.util
        if importlib.util.find_spec(package_name) is None:
            LOGGER.warn(f"Package '{package_name}' is not installed. Please install it to use this function.")
            return False
        return True
    
    required = ['heapq', 'open3d', 'collections', 'scipy', 'pathlib']
    missing = []
    errorMsg = None
    for name in required:
        if not check_and_import(name):
            missing.append(name)
            if errorMsg is None:
                errorMsg = 'To run calcChannels, please install {0}'.format(missing[0])
            else:
                errorMsg += ', ' + name

    if len(missing) > 0:
        if len(missing) > 1:
            errorMsg = ', '.join(errorMsg.split(', ')[:-1]) + ' and ' + errorMsg.split(', ')[-1]
        raise ImportError(errorMsg)

    import heapq
    import open3d as o3d
    from collections import deque
    from scipy.interpolate import CubicSpline
    from scipy.spatial import Voronoi, Delaunay
    from pathlib import Path
```

jamesmkrieger commented:
Ok. I can see the benefit of logging all the packages needed and raising one error at the end rather than going through one by one

jamesmkrieger commented:
Probably this should go in its own file not the interactions one. If anything, it’s more related to watfinder and it would be nice to have a way of comparing results and checking for waters along these channels. It would be nice to include ions too

karolamik13 commented:
All atoms that are not 'HETATM' are analyzed. Depends on which one you provide as a parsed one. If ions have 'ATOMS' instead of 'HETATM' they should be analyzed too. I can check that later and confirm.

karolamik13 commented:
James, you are right. It will fit better to waterbridges.py. 
In such a way, we would not have to create additional file.

jamesmkrieger commented:
> All atoms that are not 'HETATM' are analyzed. Depends on which one you provide as a parsed one. If ions have 'ATOMS' instead of 'HETATM' they should be analyzed too. I can check that later and confirm.

Ions are usually HETATM. We also don’t want them treated like protein atoms for defining the channels. We’d want to look at them afterwards to see if they occupy the channels.

Erykiusz commented:
Built-in visualizer can show a protein model from an stl file on top of the channels. So for example we can compute channels without ions or any HETATM atoms (which is done by default) and then visualize the computed channels with a model that includes additional atoms. This is actually an approach which I have been using for testing. Below I'm sending a screenshot showing how it looks like.
![wynik](https://github.com/user-attachments/assets/d1d6b120-2440-436d-be10-7b8339851110)

jamesmkrieger commented:
Wow! This is really pretty and informative! 

How did you make this built-in viewer? Is it based on py3dmol?

jamesmkrieger commented on prody/proteins/interactions.py:
```diff
@@ -4345,4 +4344,772 @@ def saveInteractionsPDB(self, **kwargs):
+        return
+    if not check_and_import('open3d'):
+        return
+    if not check_and_import('collections'):
```
collections is a built-in module and comes with a base installation of Python, so you don't need to check this one

Erykiusz commented:
It is based on open3d library. For it to be visualized like that it is unfortunately necessary to make the model in VMD. That's why I wrote runVmdScript function to make such models. It also needs an external script file which I'm also sending here. Github didn't let me send it as a tcl file, but for it to work it must be changed from txt to tcl.
[vmd_script.txt](https://github.com/user-attachments/files/16690768/vmd_script.txt)

jamesmkrieger commented on prody/proteins/interactions.py:
```diff
@@ -4345,4 +4344,772 @@ def saveInteractionsPDB(self, **kwargs):
+        
+        
+import logging
+import importlib.util
```
this doesn't need to imported outside the function. In fact, it could be imported inside the check_and_import function

jamesmkrieger commented on prody/proteins/interactions.py:
```diff
@@ -4345,4 +4344,772 @@ def saveInteractionsPDB(self, **kwargs):
+        
+        
+        
+import logging
```
please use prody.LOGGER instead

jamesmkrieger commented on prody/proteins/interactions.py:
```diff
@@ -4345,4 +4344,772 @@ def saveInteractionsPDB(self, **kwargs):
+    
+    def check_and_import(package_name):
+        if importlib.util.find_spec(package_name) is None:
+            LOGGER.error(f"Package '{package_name}' is not installed. Please install it to use this function.")
```
please don't use fstrings. They cause problems for building the website as we still use python 2 for that

jamesmkrieger replied:
the prody logger also has problems with LOGGER.error so please use LOGGER.warn

jamesmkrieger commented on prody/proteins/interactions.py:
```diff
@@ -4345,4 +4344,772 @@ def saveInteractionsPDB(self, **kwargs):
+    import logging
+    import importlib.util
+    
+    LOGGER = logging.getLogger(__name__)
```
again we have prody's logger with config that should be used instead

jamesmkrieger commented on prody/proteins/interactions.py:
```diff
@@ -4345,4 +4344,772 @@ def saveInteractionsPDB(self, **kwargs):
+    import importlib.util
+    
+    LOGGER = logging.getLogger(__name__)
+    logging.basicConfig(level=logging.INFO)
```
You can use confProDy(verbosity='info')

jamesmkrieger replied:
Actually you can't because of circular imports but it's better to leave the logging level to the user anyway

jamesmkrieger commented:
> It is based on open3d library. For it to be visualized like that it is unfortunately necessary to make the model in VMD. That's why I wrote runVmdScript function to make such models. It also needs an external script file which I'm also sending here. Github didn't let me send it as a tcl file, but for it to work it must be changed from txt to tcl. [vmd_script.txt](https://github.com/user-attachments/files/16690768/vmd_script.txt)

how can we need an external vmd script? Can you not have it as part of prody somewhere or have a function that writes it?

Erykiusz commented:
Ultimately, I thought it would be a file inside prody, but I didn't really know where to put it, so for now it's external.

jamesmkrieger commented:
I'd put it in prody/proteins as that's where your code is anyway. 

> Ultimately, I thought it would be a file inside prody, but I didn't really know where to put it, so for now it's external.

jamesmkrieger commented:
> James, you are right. It will fit better to waterbridges.py. In such a way, we would not have to create additional file.

Actually, I would say it is better to put this in it's own file as it is quite a big development in its own right.

I also would agree with restructuring it to have a class on the outside rather than having a function on the outside with classes and functions inside it, which means that it is essentially like a class anyway.

jamesmkrieger commented on prody/proteins/interactions.py:
```diff
@@ -4345,4 +4344,772 @@ def saveInteractionsPDB(self, **kwargs):
+        
+    print("File successfully created.")
+
+def calcChannels(atoms, r1=3, r2=1.25, min_depth=10, bottleneck=1, sparsity=15, output_path=None, visualizer=None, stl_path=None, vis_channels = None, surface=None):
```
vis_channels isn't documented

It could also be good to split these arguments across 2 lines, and put `vis_channels=None` without spaces

jamesmkrieger replied:
Is the point of this argument to visualize channels that have already been calculated rather than calculating them again? Probably it would be better to have a separate visualize function called showChannels

jamesmkrieger commented on prody/proteins/interactions.py:
```diff
@@ -4345,4 +4344,772 @@ def saveInteractionsPDB(self, **kwargs):
+        lengths = np.linalg.norm(diffs, axis=1)
+        return np.sum(lengths)
+    
+    atoms = atoms.select('not hetero')
```
It's not very easy to notice that the main function starts again properly here. If you have a class then you can define the helper functions below the main one and use self for it to know what they are

jamesmkrieger replied:
You can still have a function outside the class so that people don't need to interact with the class like we have calcANM

jamesmkrieger commented on prody/proteins/interactions.py:
```diff
@@ -4345,4 +4344,772 @@ def saveInteractionsPDB(self, **kwargs):
+    atoms = atoms.select('not hetero')
+    coords = atoms.getCoords()
+    
+    if vis_channels == None:
```
as I said above, this seems unnecessarily confusing too and you'd be better with separate calc and show functions

jamesmkrieger commented on prody/proteins/interactions.py:
```diff
@@ -4345,4 +4344,772 @@ def saveInteractionsPDB(self, **kwargs):
+    This function performs the following steps:
+    1. **Selection and Filtering:** Selects non-hetero atoms from the protein, calculates van der Waals radii, and performs
+       3D Delaunay triangulation and Voronoi tessellation on the coordinates.
+    2. **State Management:** Creates and updates different states of the protein structure to filter out simplices based on
```
what does states of the protein structure mean? This should be explained better.

My default assumption is that use mean conformational states and I guess many users would, but this is currently for a single structure.

Erykiusz replied:
States are somewhat like different steps of detecting channels. For example: there is a protein state which describes a whole protein, a surface state which describes the surface etc. States are variables with 's_' in front of them.

jamesmkrieger replied:
Ok, this definitely needs to be explained. 

I also forgot to say that you should cite the paper in the docstring too.

jamesmkrieger commented on prody/proteins/interactions.py:
```diff
@@ -4345,4 +4344,772 @@ def saveInteractionsPDB(self, **kwargs):
+    5. **Visualization and Saving:** Generates meshes for the detected channels, filters them by bottleneck size, and either
+       saves the results to a PDB file or visualizes them based on the specified parameters.
+
+    Note: Ensure that the necessary external libraries and methods are properly imported and available for this function to
```
This should probably go at the top of the docstring and with a list of what libraries are needed. Also, you don't mean `imported` as the imports will happen inside

jamesmkrieger commented on prody/proteins/interactions.py:
```diff
@@ -4345,4 +4344,772 @@ def saveInteractionsPDB(self, **kwargs):
+    from pathlib import Path
+    
+    if not check_and_import('heapq'):
+        return
```
you don't want to just return and give python the impression that everything is fine. Then it could continue to run downstream commands and have a problem

jamesmkrieger reviewed (changes requested):
The block with the imports and checks on them needs to change as it currently tries all the imports before checking them. It would also be good to actually raise an error at the end of that rather than returning None. I have provided code to do this in a comment.

I also have smaller changes like improving the documentation. There are some bigger things of restructuring but these are just suggested and not required now

jamesmkrieger commented:
> To get only selected channels we can do this: protein = prody.parsePDB(protein_path)
> 
> Calculations: channels, surface = prody.calcChannels(protein) lengths, bottlenecks = prody.getChannelsParameters(channels)
> 
> Now we got a list of lengths of the channels. Lets say we want to see only 3rd and 4th channel. So we do: selected_channels = [channels[3], channels[4]]
> 
> Now we visualize: prody.calcChannels(protein, vis_channels = selected_channels, surface = surface)

This even more shows me that it should be a separate function showChannels

> 
> I will add the option to see the atoms of channels next time I will be working on it.

Sounds good

Erykiusz commented:
Thank you for the comments and helping with code. I will try to implement the changes soon.

jamesmkrieger commented on prody/proteins/interactions.py:
```diff
@@ -4345,4 +4344,772 @@ def saveInteractionsPDB(self, **kwargs):
+        return channels, s_srf.simp
+    
+    else:
+        channel_meshes = [create_mesh_from_spline(*channel.get_splines()) for channel in vis_channels]
```
it could be good to have something like the following so that users can visualize just one channel
```
if isinstance(vis_channels, Channel):
    vis_channels = [vis_channels]
```

Otherwise, I'm having to run things like this, which is doable but a little annoying:
```
In [7]: x = prody.calcChannels(protein, vis_channels=channels[:1], surface=surface)

In [8]: x = prody.calcChannels(protein, vis_channels=channels[1:2], surface=surface)

In [9]: x = prody.calcChannels(protein, vis_channels=channels[2:3], surface=surface)
```

jamesmkrieger commented:
Have you tried any cryo-EM structures and have you tried any ion channels? I have now tried 2 cryo-EM structures of open AMPAR receptor ion channels (PDB: 8C2H and 8FP4) and the results are a bit strange, but better with the second one, I think. 

There seem to be way too many channels. I get 30 channels with default parameters and still 13 with sparsity 1000. Could this be because of missing atoms and residues?

I just tried with one where I filled in the missing parts with modeller and I think it seems better. You can find the file at https://1drv.ms/u/s!AnOJ8_qF-DUToZA715nVa5diBmfoYA?e=futY2l

Erykiusz commented:
I didn't test any cryo-EM structures, but I would suggest playing with other parameters, such as 'min_depth', which filters too small cavities. Also, 'sparsity' definies the minimum distance of surface channel points within a certain cavity, so there may be a lot of cavities detected, but they may not be really relevant. You can filter them with 'min_depth', or see if the surface was closely enough approximated with parameter 'r1' and if the channels have enough space with 'r2'.

jamesmkrieger commented:
I guess giving them more space with r2 could help for the 8c2h maybe. How did you pick the default parameters?

Erykiusz commented:
r1 and r2 were picked from MOLE, which the code is based on, min_depth was chosen as 10 to cut off very small cavities which were too many and it worked with the structure I tested, bottleneck was chosen as 1 because the channels are a bit narrower than detected cavities and 1 preserves all the channels detected with r2 as 1.25, sparsity was chosen to get the clearer results, it could be actually any positive value, the parameter is mostly to detect channels that don't overlap with each other. 

detect_channels(
protein,
r1=3,
r2=1.25,
min_depth=10,
bottleneck=1,
sparsity=15,
output_path=None,
visualizer=None,
stl_path=None,
)

jamesmkrieger commented:
Is there an option to merge channels that overlap with each other? Would I make the sparsity even lower?

Erykiusz commented:
When channels are detected they are all automatically merged to create model for the viewer and the output pdb file. If you would lower the sparcity you would get more channels that overlap, and if you would set it to 0, all the surface points of the molecule will be used as the end points of channels, resulting in many channels that highly overlap.

jamesmkrieger commented:
ok, thanks

karolamik13 commented:
> When channels are detected they are all automatically merged to create model for the viewer and the output pdb file. If you would lower the sparcity you would get more channels that overlap, and if you would set it to 0, all the surface points of the molecule will be used as the end points of channels, resulting in many channels that highly overlap.

This kind of information is very useful. At some point, we will have to include them somewhere. When the tool will be ready, I will prepare a tutorial similar to those:
http://www.bahargroup.org/prody/tutorials/
where we will have to show different cases and how different parameters change the outcome.
Later I will ask you to collect different commands somewhere in Jupyter Notebook file or so.

It would be very useful to be able to merge some selected channels (for example channel 1 and channel 9) to further compute values for them, residues that are forming the channel, etc. as for one merged channel.

Erykiusz committed:
Separate visualizer functions

Erykiusz commented:
New changes:
-runVmdScript replaced with getVmdModel, which requires only path to vmd.exe and atoms of the parsed protein. Returns protein model. (Example: model = getVmdModel(vmd_path, atoms))
-calcChannels now returns a list of channels, and a protein surface.
-Added 2 new visualizer functions: showChannels and showCavities.
	showChannels needs a list of channels to visualize, and optionally a protein model and surface. (Example: showChannels(channels, model=model, surface=surface))
	showCavities needs cavities to visualize and optionally can show protein surface with show_surface=True. (Example: showCavities(surface, show_surface=True))
-Replaced prints with LOGGER.info.
-Got rid of f-strings.
-calcChannels now has its base class with helper functions callled ChannelCalculator
-getChannelsParameters now returns also volumes of channels
-Better documentation
-Added calcChannelsMultipleFrames for computing channels from multiple frames. This is essentialy calcInteractionsMultipleFrames with slight modifications.

I didn't yet tested calcChannelsMultipleFrames with proper files so there might be errors.

@karolamik13 accessing the atoms of channels should be done to get artificial atoms making the path of the channels (hydrogen atoms) or atoms that make the surface of channels?

karolamik13 commented:
@Erykiusz "atoms that make the surface of channels" are more useful, but could we have both?

Under Lunux, vmd is usually here: 
vmd_path = '/usr/local/bin/vmd' This is the default localization, but I think we can use "whereis vmd" or another command to check it if the one above is not correct. I am not sure, though, how to do it under Windows or Mac.

I tested different functions, and they are working well except for calcChannelsMultipleFrames() where I had this error:
_---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In[23], line 1
----> 1 channels2, surfaces2 = calcChannelsMultipleFrames(p2)

File ~/anaconda3/envs/prody/lib/python3.9/site-packages/ProDy-2.5.0-py3.9-linux-x86_64.egg/prody/proteins/interactions.py:4872, in calcChannelsMultipleFrames(atoms, trajectory, output_path, **kwargs)
   4870         LOGGER.info('Model: {0}'.format(i+start_frame))
   4871         atoms.setACSIndex(i+start_frame)
-> 4872         channels = calculator.calcChannels(atoms, output_path / "_{0}.pdb".format(i+start_frame), **kwargs)
   4873         channels_all.append(channels)
   4874 else:

AttributeError: 'ChannelCalculator' object has no attribute 'calcChannels'_

I am sending multi-model PDB and trajectory via e-mail, which you can use for tests.

Erykiusz committed:
Fixed calcChannelsMultipleFrames

Erykiusz commented:
(1) Yes, this is now in progress. Both residues forming the channels and merging them are in my queue.
(2) Results can be displayed the same way as for the single pdb. To access the channels for specific frames/models we can simply write channels[n], where n is the index of a frame/model, because results are in a form of list. It is basically a list of lists. The same works for surfaces. The results can also be written in pdb form with output_path set not to none. The results are then saved as separate files for each frame.
I don't really understand what you mean with combining channels to obtain results in multi pdb/trajectories. Should it be the same as for single pdb?

karolamik13 commented:
Combining the channels: We may have two channels near each other, and we might want to join them, visualize them, and compute the volume and residues that are participating in the formation of that joint channel (=two channels or more, for example, channel[1] and channel[3]). Unless you think there is no point in doing that because each channel shows the whole pathway or for other reasons. For both single PDB and trajectory, I was thinking the same thing.

Erykiusz commented:
Ok, I will work on that

jamesmkrieger commented:
I guess there should be some criteria that are different for combining channels in a trajectory e.g. if they are close in one frame then they could be combined in all frames. Maybe we end up with a channel that has two disconnected parts in some frames and that would be interesting to see.

jamesmkrieger commented:
It could be nice to have a way to turn channels on and off in the viewer, so we could open a few of them and then focus on particular ones. It would also be great if we can open a few frames and turn them on and off to have everything in the same viewer to cross compare rather than having to open and close the viewers in between or open multiple instances. 

Is it possible to continue typing in the terminal while a viewer is open to perform further analysis based on what we see and open other viewers?

Also, what quantitative measures do we have like width along the channel with some connection to lining residues and can we compare these across frames? Could we maybe make a plot of width near a particular residue over time for example?

Erykiusz committed:
added getChannelAtoms

Erykiusz commented:
I've added getChannelAtoms which returns an AtomGroup object of dummy atoms forming the channels. The user can also choose a 'resolution' which is basically a number of samples multipled by the length of the channel. The default value is 5 which is the same as for outputing pdb file. When calling getChannelAtoms we get an object that merges all the channels into one.
Example usage: channel_atoms = getChannelAtoms(channels)

Also, I don't really see the need for merging channel because the values from getChannelParameters and getChannelParameters are from the list of channels anyway. If we would want to get values for specific channels we could just simply manipulate elements of the channel list. For example:
selected_channels = channels[1:4]

And then we could get the parameters and atoms making all the channels with:
lengths, bottlenecks, volumes = getChannelParameters(selected_channels)
selected_channels_atoms = getChannelAtoms(selected_channels)

We could also visualize only the channels of interest:
showChannels(selected_channels, model)

Another thing is that channels are saved as CubicSplines and merging them would result in very strange results. If the merging of channels would be necessery, I would think of a solution, but I think the method I described above should be enough, but correct me if I'm wrong.

I also changed getChannelsParameters to getChannelParameters.

When it comes to viewer, I didn't experiment with changing it while it is running and I don't know if it is possible. Opening multiple viewers is not possible right now.

We can compare the quantitive measures from getChannelsParameters from different frames and for now it is an only option.

Erykiusz committed:
small fixes

jamesmkrieger commented:
Ok, makes sense. Thanks

karolamik13 commented:
I tested everything, and it works great!

Two suggestions:
(1) Can we add a function that will save all detected channels separately instead of together too?
**channels, surface = calcChannels(atoms, output_path='channels_output.pdb')**
for example by applying 'output_all'='channel' and then we will obtain channel_1.pdb, channel_2.pdb, etc.

(2) Can you create a function like getChannelAtoms() that will return protein + channel coordinates (or add a parameter to include it in getChannelAtoms())?
If you have protein+channel's atoms, you can use such command:
distance = 4
residues = **prot_and_channels_atoms**.select('same residue as exwithin '+str(distance)+' of resname FIL')
and you will have information about residues (from protein structure) that are participating in the formation of the channel. 
The last thing that is needed.

Erykiusz committed:
protein + channels and separate outputs for channels

Erykiusz commented:
I've added a parameter 'separate' to calcChannels and calcChannelsMultipleFrames that enables writing separate output files for each channel. Examples:
channels, surface = calcChannels(protein, output, separate=True)
all_channels, all_surfaces=calcChannelsMultipleFrames(multi, traj, output_path = output, separate=True)

Also getChannelAtoms now has a parameter 'protein' that when set to an AtomGroup object (parsed protein) will merge itsef with the channels. Example:
atoms_protein = getChannelAtoms(channels, protein)

Erykiusz committed:
fixing documentation

karolamik13 commented:
Works great. We can also add a function to obtain residues that are forming the channel, but it can be done later. 
Now everyone gets this by using:
**p = parsePDB('1tqn.pdb')
atoms = p.select("protein")
channels, surface = calcChannels(atoms)**

# for channel[1]:
**atoms_protein = getChannelAtoms(channels[1], atoms)
distance = 4
residues = atoms_protein.select('same residue as exwithin '+str(distance)+' of resname FIL')
list(zip(residues.select('name CA').getResnames(), residues.select('name CA').getResnums()))**

Once the checks are done successfully (at least for Python 3.8 and higher), I will accept the pull request.

jamesmkrieger commented:
Ideally we should remove the fstrings from interactions.py so that the python 2 import works too

karolamik13 commented:
> Ideally we should remove the fstrings from interactions.py so that the python 2 import works too

Yes, you are right.

Erykiusz committed:
f-string removed

Erykiusz commented:
Are the f-strings in my code causing problems? Because there was only one f-string in checkAndImport, which I removed.

karolamik13 commented:
If you look here:
https://github.com/karolamik13/ProDy/actions/runs/10585767777/job/29333210140?pr=7

You will see that also that line is problematic:
s_tmp.set_state(*calculator.delete_simplices3d(coords, *s_tmp.get_state(), vdw_radii, r1, True))

and probably would work if you make it like the one below so that it would be compatible with Python 2.7.
state = calculator.delete_simplices3d(coords, *s_tmp.get_state())
s_tmp.set_state(state, vdw_radii, r1, True)

But we might have more of those in a code. It works for Python 3.8 and higher, so I am happy with the code.

@jamesmkrieger maybe it would be a good time to finally drop Python 2.7? 
If not then we will look later at that to fix it.

jamesmkrieger commented:
We would have to change how we build the website, but yes, it would be great if we could
